bpf-gpl/tc.c

```diff
@@ -922,31 +931,53 @@ static CALI_BPF_INLINE struct fwd calico_tc_skb_accepted(struct __sk_buff *skb,
 		CALI_DEBUG("CT: DNAT to %x:%d\n",
 				be32_to_host(state->post_nat_ip_dst), state->post_nat_dport);
 
-		struct cali_rt *rt;
 
 		encap_needed = dnat_should_encap();
-		if (encap_needed) {
-			rt = cali_rt_lookup(state->post_nat_ip_dst);
-			if (!rt) {
-				reason = CALI_REASON_RT_UNKNOWN;
-				goto deny;
-			}
-			CALI_DEBUG("rt found for 0x%x local %d\n",
-					be32_to_host(state->post_nat_ip_dst), !!cali_rt_is_local(rt));
-
-			encap_needed = !cali_rt_is_local(rt);
-		}
 
 		/* We have not created the conntrack yet since we did not know
 		 * if we need encap or not. Must do before MTU check and before
 		 * we jump to do the encap.
 		 */
 		if (ct_rc == CALI_CT_NEW) {
-			if (CALI_F_DSR && CALI_F_FROM_HEP &&
-					encap_needed && state->nat_tun_src == 0) {
-				ct_nat_ctx.flags |= CALI_CT_FLAG_DSR_FWD;
+			struct cali_rt * rt;
+			int nat_type = CT_CREATE_NAT;
+
+			if (encap_needed) {
+				rt = cali_rt_lookup(state->post_nat_ip_dst);
+				if (!rt) {
+					reason = CALI_REASON_RT_UNKNOWN;
+					goto deny;
+				}
+				CALI_DEBUG("rt found for 0x%x local %d\n",
+						be32_to_host(state->post_nat_ip_dst), !!cali_rt_is_local(rt));
+
+				encap_needed = !cali_rt_is_local(rt);
+			}
+
+			if (encap_needed) {
```

Comment:
Duplicate condition?  Could these two blocks be combined?

Reply:
The first block may change the condition for the second, but we can merge them, There was previously more code in-between. Done.